### PR TITLE
fetchdarcs: add SSL_CERT_FILE environment variable

### PR DESCRIPTION
**Pull request by k0001**
This is the same as #25250 but against `release-1703`

cc @shlevy @luite

### CODE REVIEW

File: pkgs/build-support/fetchdarcs/default.nix [hunk 1]
```diff
@@ -1,10 +1,13 @@
...
+{stdenv, darcs, nix, cacert}:
+
+{url, rev ? null, context ? null, md5 ? "", sha256 ? ""}:
 
 if md5 != "" then
   throw "fetchdarcs does not support md5 anymore, please use sha256"
 else
 stdenv.mkDerivation {
   name = "fetchdarcs";
+  SSL_CERT_FILE = "${cacert}/etc/ssl/certs/ca-bundle.crt";
```
Comment (copumpkin) [inline pkgs/build-support/fetchdarcs/default.nix]: Shouldn't this be `NIX_SSL_CERT_FILE` nowadays? @edolstra is there central guidance on how the variable should be used across the repo?

Comment (k0001) [inline pkgs/build-support/fetchdarcs/default.nix]: `SSL_CERT_FILE` works. I can change this to `NIX_SSL_CERT_FILE` if that's the recommendation and works all the same.

**Comment by copumpkin** [comment]
I think the idea is that `NIX_SSL_CERT_FILE` _should_ work, and if it
doesn't we probably should make it. Probably not something for this PR but
I'd leave a comment in the code that it's weird

**Comment by shlevy** [comment]
`NIX_SSL_CERT_FILE` is patched into our `openssl`, but its purpose is to avoid clashing with host tools using host certificate chain. Seems irrelevant for within builds.

File: pkgs/build-support/fetchdarcs/default.nix (hunk 1 above)
Comment (k0001) [inline pkgs/build-support/fetchdarcs/default.nix]: FWIW, I just copied over this usage of `SSL_CERT_FILE` from `fetchbower`.

**Comment by copumpkin** [comment]
@shlevy just saying because it's unclear. For example I think we amended our patch to the Go http library so that it looks for `NIX_SSL_CERT_FILE` _instead of_ `SSL_CERT_FILE` (which we had to patch in previously). So for tools that we expect to use in builds _and_ as user tools, that's where this sort of question becomes more important, or at least it's unclear to me. If someone wrote some sort of go downloader fixed output derivation, they'd need to pass in `NIX_SSL_CERT_FILE` for it to work properly.

**Comment by shlevy** [comment]
Oh, I don't think it's correct to do "instead" semantics. `NIX_SSL_CERT_FILE` should be an override.

**Comment by kirelagin** [comment]
@copumpkin @shlevy For me this change looks somewhat confusing too.

Am I right in my understading that the idea is to make sure that TLS verifications _during the build_ (e.g. fetching sources, etc.) use a fixed set of trusted certs (the one from `pkgs.cacert`), while making sure at the same time that the path to this bundle is never embedded in the binaries, so that during runtime binaries do their best to use a system-default location (or `NIX_SSL_CERT_FILE` as an override)?

Aren’t similar changes necessary for all the `fetch*` functions out there? For example `fetchgit` does not seem to override anything, which will probably result in a (minor) impurity: if my system trusted roots are different from Hydra’s it is possible that the build will fail for me locally (won’t be able to fetch surces) while succeeding on Hydra. Is that what we want to avoid?

Or could it be that we want the opposite: the system’s preferred certificates to be always used, even during the build?

**Comment by kirelagin** [comment]
I stand corrected, `fetchgit` contains the override. But, looks like, e.g. `fetchurl` does not.

**Comment by edolstra** [comment]
Why does fetchdarcs need certificates at all? `fetchurl` doesn't need them because it uses `curl --insecure`. This is fine because the integrity of the downloaded file is guaranteed by the hash.

**Comment by kirelagin** [comment]
@edolstra Ah, yes, of course. Since we are checking hashes anyway, all of the `fetch*` functions should probably get rid of the `pkgs.cacert` dependency and just somehow force unsecure download.

**Comment by kirelagin** [comment]
#26273

**Comment by shlevy** [comment]
The integrity of the downloaded file is guaranteed, sure, but the privacy of what is being requested isn't!

**Pull request merged**